homeassistant/components/vicare/binary_sensor.py

```diff
@@ -0,0 +1,123 @@
+"""Viessmann ViCare sensor device."""
+import logging
+
+import requests
+
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASS_POWER,
+    BinarySensorDevice,
+)
+from homeassistant.const import CONF_DEVICE_CLASS, CONF_NAME
+
+from . import (
+    DOMAIN as VICARE_DOMAIN,
+    PYVICARE_ERROR,
+    VICARE_API,
+    VICARE_HEATING_TYPE,
+    VICARE_NAME,
+    HeatingType,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_GETTER = "getter"
+
+SENSOR_CIRCULATION_PUMP_ACTIVE = "circulationpump_active"
+SENSOR_BURNER_ACTIVE = "burner_active"
+SENSOR_COMPRESSOR_ACTIVE = "compressor_active"
+
+SENSOR_TYPES = {
+    SENSOR_CIRCULATION_PUMP_ACTIVE: {
+        CONF_NAME: "Circulation pump active",
+        CONF_DEVICE_CLASS: DEVICE_CLASS_POWER,
+        CONF_GETTER: lambda api: api.getCirculationPumpActive(),
+    },
+    SENSOR_BURNER_ACTIVE: {
+        CONF_NAME: "Burner active",
+        CONF_DEVICE_CLASS: DEVICE_CLASS_POWER,
+        CONF_GETTER: lambda api: api.getBurnerActive(),
+    },
+    SENSOR_COMPRESSOR_ACTIVE: {
+        CONF_NAME: "Compressor active",
+        CONF_DEVICE_CLASS: DEVICE_CLASS_POWER,
+        CONF_GETTER: lambda api: api.getCompressorActive(),
+    },
+}
+
+SENSORS_GENERIC = [SENSOR_CIRCULATION_PUMP_ACTIVE]
+
+SENSORS_BY_HEATINGTYPE = {
+    HeatingType.gas: [SENSOR_BURNER_ACTIVE],
+    HeatingType.heatpump: [SENSOR_COMPRESSOR_ACTIVE],
+}
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Create the ViCare sensor devices."""
+    if discovery_info is None:
+        return
+
+    vicare_api = hass.data[VICARE_DOMAIN][VICARE_API]
+    heating_type = hass.data[VICARE_DOMAIN][VICARE_HEATING_TYPE]
+
+    sensors = SENSORS_GENERIC
+
+    if heating_type != HeatingType.generic:
+        sensors.extend(SENSORS_BY_HEATINGTYPE[heating_type])
```

Comment:
We're not allowed to modify the constant `sensors`. Either copy the constant or split the iteration.

Reply:
I think it's simpler if just iterate all sensor types and use `continue` in a check.
```py
SENSORS_BY_HEATINGTYPE = {
    HeatingType.generic: SENSOR_CIRCULATION_PUMP_ACTIVE,
    HeatingType.gas: SENSOR_BURNER_ACTIVE,
    HeatingType.heatpump: SENSOR_COMPRESSOR_ACTIVE,
}


def setup_platform(hass, config, add_entities, discovery_info=None):
    """Create the ViCare sensor devices."""
    if discovery_info is None:
        return

    vicare_api = hass.data[VICARE_DOMAIN][VICARE_API]
    heating_type = hass.data[VICARE_DOMAIN][VICARE_HEATING_TYPE]
    name = hass.data[VICARE_DOMAIN][VICARE_NAME]

	for sensor in SENSOR_TYPES:
        sensor_type = SENSORS_BY_HEATINGTYPE[heating_type]
	    if sensor != sensor_type and sensor != SENSOR_CIRCULATION_PUMP_ACTIVE:
	        continue
	    
	    entity = ViCareBinarySensor(name, vicare_api, sensor)
	    entities.append(entity)

	add_entities(entities)
```

Reply:
I've decided to go for the copy method. I think it looks cleaner if you consider that you need to repeat all generic sensors in the loop in sensor.py.